### PR TITLE
Add edit modal and delete user

### PR DESCRIPTION
**Pull request by barbara-chaves**
### General description

This adds the 'edit' button to the user list, with edit the form disabled and the 'delete user' functionality working
It changes the users list to don't display the current user name

It updates the users page style and table content 

### Designs

[users list](https://www.figma.com/file/7bmF29CIXDJwCpuloW5Nt6/%F0%9F%94%B5-Landgriffon---Redesign_BLUE?type=design&node-id=1870-167404&mode=dev)
[edit user modal](https://www.figma.com/file/7bmF29CIXDJwCpuloW5Nt6/%F0%9F%94%B5-Landgriffon---Redesign_BLUE?type=design&node-id=1870-169568&mode=dev)

### Testing instructions

1. With an admin account, go to '/profile/users'
2. the list should NOT display your current user name
3. click on some 'edit' button on the list. A modal should open with the edit user form and the 'save' button should be disabled
4. At the bottom of the list are some test users to be deleted. Try to delete them.

### Tasks
https://vizzuality.atlassian.net/browse/LANDGRIF-1346
https://vizzuality.atlassian.net/browse/LANDGRIF-1386

### CODE REVIEW

File: client/src/containers/delete-user/component.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  const [open, setOpen] = useState(false);
+  const { mutate: deleteUser, isLoading } = useDeleteUser();
+
+  const handleDeleteUser = () => {
+    deleteUser(id, {
+      onSettled: () => setOpen(false),
+      onError: () => toast.error('There was an error deleting the user. Try again later.'),
+      onSuccess: () => toast.success('User deleted successfully'),
+    });
+  };
```
Comment (davidsingal) [inline client/src/containers/delete-user/component.tsx:24]: here we could use `useCallback`

File: client/src/containers/delete-user/component.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  return (
+    <>
+      <Button size="xs" variant="secondary" danger onClick={() => setOpen(true)}>
+        Delete user
+      </Button>
+      <Modal
+        theme="minimal"
+        title={`Delete user "${userName}"`}
+        open={open}
+        onDismiss={() => setOpen(false)}
```
Comment (davidsingal) [inline client/src/containers/delete-user/component.tsx:35]: here also, `useCallback`

File: client/src/containers/delete-user/component.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+          <div>
+            <h1 className="text-base leading-7 font-semibold text-gray-900 mb-1">
+              Delete user &quot;{userName}&quot;
+            </h1>
+            <p className="text-sm text-gray-500 mb-4">
+              All of this user data will be permanently removed from our servers forever. This
+              action cannot be undone.
+            </p>
+            <div className="flex justify-end gap-4">
+              <Button variant="white" onClick={() => setOpen(false)}>
```
Comment (davidsingal) [inline client/src/containers/delete-user/component.tsx:52]: reuse `onDismiss` function you used above

File: client/src/containers/edit-user/component.tsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+  return (
+    <>
+      <Button size="xs" variant="white" className="!text-sm" onClick={() => setOpenModal(true)}>
+        Edit
+      </Button>
+      <Modal open={open} title={`Edit user ${userName}`} onDismiss={() => setOpenModal(false)}>
+        <UserForm user={user}>
+          <div className="flex justify-between flex-1 mr-2.5 px-0.5">
+            <DeleteUser id={user.id} userName={userName} />
```
Comment (davidsingal) [inline client/src/containers/edit-user/component.tsx:29]:
```suggestion
            <DeleteUser data={user} />
```
I'd prefer to pass the user object, and `userName` should be created inside `DeleteUser` component

File: client/src/containers/delete-user/component.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import { Button } from 'components/button';
+import Modal from 'components/modal/component';
+import { useDeleteUser } from 'hooks/users';
+
+type DeleteUserProps = {
+  userName: string;
+  id: string;
+};
+
+const DeleteUser = ({ id, userName }: DeleteUserProps) => {
```
Comment (davidsingal) [inline client/src/containers/delete-user/component.tsx:14]:
```suggestion
const DeleteUserModal = ({ id, userName }: DeleteUserProps) => {
```
I'd prefer to add more info in the component naming.

File: client/src/containers/edit-user/user-form.tsx [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    defaultValues: {
+      ...pick(user, 'fname', 'lname', 'displayName', 'email'),
+      roles: user.roles.map((role) => role.name),
+    },
+  });
+
+  const roleOptions = useMemo(
+    () => Object.values(RoleName).map((role) => ({ label: role, value: role })),
+    [],
+  );
```
Comment (davidsingal) [inline client/src/containers/edit-user/user-form.tsx:33]: `useMemo` is to much for an array of only 2 items, you could remove it

File: client/src/containers/edit-user/user-form.tsx [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+  const [roleValue, setRoleValue] = useState<Option<string>[]>(
+    user.roles.map((role) => ({ label: role.name, value: role.name })),
+  );
```
Comment (davidsingal) [inline client/src/containers/edit-user/user-form.tsx:37]:
```suggestion
  const [roleValue, setRoleValue] = useState<Option<string>[]>(roleOptions);
```
Couldn't be like this?

File: client/src/types.d.ts [hunk 8]
```diff
@@ -362,10 +362,10 @@ export type Task = {
 // User
 export type User = {
   displayName: string;
```
Comment (davidsingal) [inline client/src/types.d.ts:364]: `displayName` will be removed in the API

**Review by davidsingal — CHANGES_REQUESTED** [review]

File: client/src/containers/edit-user/component.tsx (hunk 4 above)
Comment (barbara-chaves) [reply client/src/containers/edit-user/component.tsx:29]: I created the useName outside because it is used by the parent element too

File: client/src/containers/edit-user/user-form.tsx (hunk 7 above)
Comment (barbara-chaves) [reply client/src/containers/edit-user/user-form.tsx:37]: It can’t because the  initial state is related to the user data coming from the API, and the roleOptions is a hardcoded value

File: client/src/types.d.ts (hunk 8 above)
Comment (barbara-chaves) [reply client/src/types.d.ts]: I will remove this property and replace it on components by the user full name

File: client/src/utils/user-full-name.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import type { User } from 'types';
+
+const getUserFullName = (user: User, replaceByEmail = false) => {
```
Comment (andresgnlez) [inline client/src/utils/user-full-name.ts:3]: Instead of a `replaceByEmail = false` parmeter, I would suggest something like an `options` object. Why? Well, if in the future there are more options to contemplate, with the current approach, you will need to add more parameters to the function, meaning updating the function where is used and forcing to specific all values eventually.

If you have something like 
```
options = {
  replaceByEmail: false,
  useOnlyFirstName: true,
  someRandomFutureOption: false,
},

// your logic
if (options.replaceByEmail) //...
if (options.useOnlyFirstName) //...
if (options.someRandomFutureOption) //...
```

This way you only need to work extending the business logic of the function and forget about updating parameters across the codebase.

TLDR: the code si completely fine, it's just a matter of scalability

Comment (andresgnlez) [reply client/src/utils/user-full-name.ts:3]: Also, don't forget to type the return of the function with `:string` :)

File: client/src/utils/user-full-name.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+  if (!user) return '';
+  const { fname, lname, email } = user;
+  if (fname) {
+    return `${fname}${` ${lname}` || ''}`;
```
Comment (andresgnlez) [inline client/src/utils/user-full-name.ts:7]: Isn't it simpler just ``${fname} ${lname}`` (notice the space between the two variables) ? Or am I missing something? 🤔

File: client/src/containers/user-dropdown/component.tsx [hunk 11]
```diff
@@ -23,6 +24,8 @@ const UserDropdown: React.FC = () => {
 
   const handleSignOut = useCallback(() => signOut({ callbackUrl: '/auth/signin' }), []);
 
+  const userName = useMemo(() => getUserFullName(user), [user]);
```
Comment (andresgnlez) [inline client/src/containers/user-dropdown/component.tsx:27]: No need of using `useMemo` here, the operation is not that expensive to optimize it.

File: client/src/utils/user-full-name.ts (hunk 9 above)
Comment (andresgnlez) [reply client/src/utils/user-full-name.ts:3]: Also (and I finish, for real), I think is more useful for you, and other developers, knowing the API of your function is `getUserFullName(user, { replaceByEmail: true })` instead of `getUserFullName(user, true)` (as you don't what that `true` really means and would need to check on the function)

**Comment by andresgnlez** [comment]
image
@barbara-chaves  I tried to update my info but the _save_ button is disabled, is this expected? Also, can a user have the roles _admin_ and _user_ at the same time? Sounds weird to me 🤔

File: client/src/utils/user-full-name.ts (hunk 9 above)
Comment (barbara-chaves) [reply client/src/utils/user-full-name.ts:3]: Great review! Thanks!

File: client/src/utils/user-full-name.ts (hunk 10 above)
Comment (barbara-chaves) [reply client/src/utils/user-full-name.ts:7]: I cant use `${fname} ${lname}` because it can return a 'undefined undefined' for example, but my logic isn't good either. I will improve the function.

Comment (andresgnlez) [reply client/src/utils/user-full-name.ts]: You can check in the `if` if both variables exist, if so, print both names, otherwise, it will print what you have in the final `return` of the function (ideally `fname`)

**Review by andresgnlez — APPROVED** [review]
